### PR TITLE
Tsification binge #7

### PR DESCRIPTION
**Pull request by OfekShilon**

### CODE REVIEW

**Comment by partouf** [comment]
frontend test fail looks like a real bug

**Comment by OfekShilon** [comment]
@partouf  fixed

File: lib/compilers/cerberus.ts [hunk 1]
```diff
@@ -114,10 +115,10 @@ export class CerberusCompiler extends BaseCompiler {
         }
     }
 
-    override async processAsm(result) {
+    override async processAsm(result): Promise<ParsedAsmResult> {
         // Handle "error" documents.
         if (!result.asm.includes('\n') && result.asm[0] === '<') {
-            return [{text: result.asm, source: null}];
+            return {asm: [{text: result.asm, source: null}]};
```
Comment (OfekShilon) [inline lib/compilers/cerberus.ts:121]: This is a real semantic fix

File: lib/compilers/d8.ts [hunk 2]
```diff
@@ -220,9 +220,9 @@ export class D8Compiler extends BaseCompiler implements SimpleOutputFilenameComp
     }
 
     // Map line numbers to lines.
-    override async processAsm(result) {
+    override async processAsm(result): Promise<ParsedAsmResult> {
         if (result.code !== 0) {
-            return [{text: result.asm, source: null}];
+            return {asm: [{text: result.asm, source: null}]};
```
Comment (OfekShilon) [inline lib/compilers/d8.ts:225]: Same fix

File: lib/options-handler.ts [hunk 3]
```diff
@@ -56,6 +56,7 @@ export type VersionInfo = {
     options: string[];
     hidden: boolean;
     packagedheaders?: boolean;
+    $order?: number;
```
Comment (OfekShilon) [inline lib/options-handler.ts:59]: This is used in https://github.com/compiler-explorer/compiler-explorer/blob/9e050befc6025e4bc37a4a95680cf6bc03c1fe8c/lib/options-handler.ts#L369 , should be a member of VersionInfo.

File: static/assert.ts [hunk 4]
```diff
@@ -67,7 +67,7 @@ function fail(fail_message: string, user_message: string | undefined, args: any[
     }
 }
 
-export function assert<C>(c: C, message?: string, ...extra_info: any[]): asserts c {
```
Comment (OfekShilon) [inline static/assert.ts:70]: This is a TS bug, discussed here: https://github.com/microsoft/TypeScript/issues/60130

File: static/panes/diff.ts [hunk 5]
```diff
@@ -90,9 +90,9 @@ class DiffStateObject {
         this.extraoption = extraoption;
     }
 
-    update(id: number | string, compiler, result: CompilationResult) {
+    update(id: number | string, compiler: CompilerInfo, result: CompilationResult) {
         if (this.id !== id) return false;
-        this.compiler = compiler;
+        this.compiler!.compiler = compiler;
```
Comment (OfekShilon) [inline static/panes/diff.ts:95]: A very real fix

File: static/panes/diff.ts [hunk 6]
```diff
@@ -320,7 +320,7 @@ export class Diff extends MonacoPane<monaco.editor.IStandaloneDiffEditor, DiffSt
         this.updateCompilers();
     }
 
-    getDiffableOptions(picker?, extraoptions?: DiffOption[]): any[] {
+    getDiffableOptions(picker?: HTMLSelectElement | TomSelect, extraoptions?: DiffOption[]): any[] {
```
Comment (OfekShilon) [inline static/panes/diff.ts:323]: Just documents the weird current flow..

File: static/panes/editor.ts [hunk 7]
```diff
@@ -498,7 +498,7 @@ export class Editor extends MonacoPane<monaco.editor.IStandaloneCodeEditor, Edit
         super.initializeGlobalDependentProperties();
 
         this.httpRoot = window.httpRoot;
-        this.langKeys = Object.keys(languages);
+        this.langKeys = Object.keys(languages) as LanguageKey[]; // Why, ts. Why.
```
Comment (OfekShilon) [inline static/panes/editor.ts:501]: `Object.keys` of a dictionary are untyped, even if the dictionary is

Comment (junlarsen) [reply static/panes/editor.ts:501]: There's actually a sane reason for this: https://alexharri.com/blog/typescript-structural-typing

**Comment by OfekShilon** [comment]
Had to rebase and force push - as time goes by such PRs are getting harder to keep in sync.
I'll go ahead and merge.

**Pull request merged**